mypy/checkexpr.py

```diff
@@ -1335,6 +1345,46 @@ def check_callable_call(
             callee = callee.copy_modified(ret_type=new_ret_type)
         return callee.ret_type, callee
 
+    def _check_non_ret_from_prot(self, type: TypeInfo, attr_name: str) -> bool:
+        """Determine whether the attribute is a function with None-compatible return type.
+
+        Additionally, check whether the function was defined in a protocol.
+        """
+        if not state.strict_optional:
+            # If strict-optional is not set, is_subtype(NoneType(), T) is always True.
+            # So, we cannot do anything useful here in that case.
+            return False
+        for base in type.mro:
+            symnode = base.names.get(attr_name)
+            if symnode is None or not base.is_protocol:
+                continue
+            node = symnode.node
+            if isinstance(node, OverloadedFuncDef):
+                if node.impl is not None:
```

Comment:
As I said above, overloads without implementation are explicitly abstract. So this whole block can be turned into `node = node.impl`